### ISSUE TITLE
PETSc does not compile / update to FEniCS 2019.1

### ISSUE THREAD

**Issue by asceenl** [sigvaldm/arch-fenics-packages#1]
When installing PETSc, the compilation stalls at

```
/home/anne/src/arch-fenics-packages/petsc/src/petsc-3.9.2/src/sys/objects/init.c: In function ‘PetscOptionsCheckInitial_Private’:                            
/home/anne/src/arch-fenics-packages/petsc/src/petsc-3.9.2/src/sys/objects/init.c:420:40: error: ‘MPI_Handler_function’ undeclared (first use in this function
); did you mean ‘MPI_User_function’?                                                                                                                         
     ierr = MPI_Comm_create_errhandler((MPI_Handler_function*)Petsc_MPI_DebuggerOnError,&err_handler);CHKERRQ(ierr);                                         
                                        ^~~~~~~~~~~~~~~~~~~~                                                                                                 
                                        MPI_User_function                                                                                                    
/home/anne/src/arch-fenics-packages/petsc/src/petsc-3.9.2/src/sys/objects/init.c:420:40: note: each undeclared identifier is reported only once for each func
tion it appears in
/home/anne/src/arch-fenics-packages/petsc/src/petsc-3.9.2/src/sys/objects/init.c:420:61: error: expected expression before ‘)’ token
     ierr = MPI_Comm_create_errhandler((MPI_Handler_function*)Petsc_MPI_DebuggerOnError,&err_handler);CHKERRQ(ierr);
                                                             ^
/home/anne/src/arch-fenics-packages/petsc/src/petsc-3.9.2/src/sys/objects/init.c:420:12: error: too few arguments to function ‘MPI_Comm_create_errhandler’
     ierr = MPI_Comm_create_errhandler((MPI_Handler_function*)Petsc_MPI_DebuggerOnError,&err_handler);CHKERRQ(ierr);
            ^~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by amdj** [comment]
I think it turns out that openmpi should also be version-fixed. I am working on this.

**Comment by asceenl** [comment]
It turns out that openmpi should also be version-fixed. I am working on this.

**Comment by sigvaldm** [comment]
Right. It could also be that we could actually switch to a newer PETSc now, in which case it would be preferrable to update the PETSc in the repo. If not then I suppose we need to lock the version of OpenMPI. Let me know what you figure out.

**Comment by asceenl** [comment]
In the past I did something similar as you: trying to create a stack of version-fixed packages for FEniCs / Dolfin. In that work, I found that Dolfin 2018.x. does not compile on PETSc 3.10. A fix is present in a newer Dolfin version, however I do not know whether a stable version already incorporates this fix.

**Comment by asceenl** [comment]
I got everything compiled and running with openmpi-3.1.3-1 build, as of 2019-04-11. However I added the repo with `asp checkout`, so you have to `cd` into `openmp/repos/extra-x86_64` before running `makepkg -is`. I forked this repo and pushed it to a branch. If you would like to add this, you can pull in the changes.

**Comment by sigvaldm** [comment]
I had a look at your branch openmpi3, but I think something went wrong somewhere, because the git submodule `openmpi` does not work:

```
$ git submodule init 
Submodule 'openmpi' (ssh://github.com/sigvaldm/arch-fenics-packages/openmpi) registered for path './'
$ git submodule update 
Cloning into '/home/sigvald/Codes/arch-fenics-packages/openmpi'...
fatal: remote error: 
  sigvaldm/arch-fenics-packages/openmpi is not a valid repository name
  Email support@github.com for help
fatal: clone of 'ssh://github.com/sigvaldm/arch-fenics-packages/openmpi' into submodule path '/home/sigvald/Codes/arch-fenics-packages/openmpi' failed
Failed to clone 'openmpi'. Retry scheduled
Cloning into '/home/sigvald/Codes/arch-fenics-packages/openmpi'...
fatal: remote error: 
  sigvaldm/arch-fenics-packages/openmpi is not a valid repository name
  Email support@github.com for help
fatal: clone of 'ssh://github.com/sigvaldm/arch-fenics-packages/openmpi' into submodule path '/home/sigvald/Codes/arch-fenics-packages/openmpi' failed
Failed to clone 'openmpi' a second time, aborting
```

Indeed, when I open `.gitmodules`, the `url` doesn't make sense:

```
[submodule "openmpi"]
	path = openmpi
	url = ./openmpi
```

It would be nice to have your OpenMPI PKGBUILD, because you're probably not the only one having trouble with OpenMPI.

And what is `asp checkout` btw?

**Comment by asceenl** [comment]
Yes I saw. Some stuff with submodules I still had to learn. `asp` is the Arch build system tool that uses svntogit. I created a new copy of the PKGBUILD tree for `openmpi`, as it is part of the `extra` repo of Arch. I pushed it to my own Github and set the right link to the repo in de `openmpi3` branch. Can you check that it is working right now?

**Comment by sigvaldm** [comment]
I don't have time today, but I'll try to have a look at it during the weekend or next week. Don't hesitate to remind me if I forget.

**Comment by g-voelker** [comment]
I compiled dolfin 2019.1.0.post0 successfully with

python-sympy 1.4.1
openmpi 4.0.1-1
petsc 3.11.1-1
slepc 3.11.1-1
python-ufl 2019.1.0
python-dijitso 2019.1.0
python-fiat 2019.1.0
python-ffc 2019.1.0

It's cumbersome to change all the check sums but it works. If you want I can create a pull request. 
Cheers

**Comment by sigvaldm** [comment]
Now that FEniCS 2019 has arrived (it wasn't when this thread was opened), it's definitely better to upgrade FEniCS than to downgrade PETSc, so I'm interested in your solution @g-voelker.

I suppose you changed the PKGBUILD files of python-ffc, python-fiat, and so on? If so we would have to push them to the Arch User Repository first, since this repository only includes them as submodules, and then update this repository.

By the way, what's the thing with `.post0`? I've never used it before.

**Comment by g-voelker** [comment]
@sigvaldm That's exactly what I have done. I have just forked this repo but will not have the time to pack and update everything right now. I am not sure I can push changes to the AUR directly, can I? There is a read-only git repo. Since you are the package owner I suppose you've got access to them. How would it be convenient for you?

Concerning the weird version: There was a bug in a commit for petsc 3.11 compatibility in 2019.1.0 which was fixed some days after the release so that they pushed a post0 version to their releases. I just adapted the release version to their nomenclature. You can see their commits here:
https://bitbucket.org/fenics-project/dolfin/commits/all

**Comment by sigvaldm** [comment]
If you have time to do it later, and you have a username on the Arch User Repository, I can add you as a co-maintainer of the FEniCS-packages. That should grant you write-access to the repositories.

Once that is done I'd like to update all the submodules in this repository to point to the newest version of those AUR repositories, as well as the commit of the PETSc/SLEPc/OpenMPI AUR repositories that you used. Since it works with SymPy 1.4, now, we should remove `python-sympy-sm`, and update the README. 

What do you think?

**Comment by g-voelker** [comment]
That sounds like a plan. My AUR username is "Polly".

SLEPc will need to be added to AUR (the submodule link is broken with a 404)
For the openmpi we can point to the version in extra.

**Comment by sigvaldm** [comment]
Right, sorry, we don't have to point to OpenMPI since it's in extra and not in AUR (unless you couldn't just use the newest version). I guess it will be installed when installing hdf5-openmpi as described in the README.

I added you to all FEniCS packages. Remember to reset the `pkgrel` when you increase the FEniCS-version and add yourself as co-maintainer in the comments.

I checked the SLEPc package, and it seems it has been orphaned. I adopted it, and added you there as well. More to keep track of for us, but at least it makes it easier to sync all the versions on AUR.

**Comment by g-voelker** [comment]
I have updated the AUR packages. Updating the submodules in this repo should fix the issue.

**Comment by sigvaldm** [comment]
Thank you. Would you like to do it and send me a PR or should I do it? (includes removing `python-sympy-sm` and move sympy to the `pacman` command in the README)

**Comment by g-voelker** [comment]
I am just on the update. Is there a reason to stay with scotch rather than then AUR package (https://aur.archlinux.org/packages/scotch/)?

**Comment by g-voelker** [comment]
Same question about parmetis (https://aur.archlinux.org/packages/parmetis/)

**Comment by g-voelker** [comment]
In my fork I moved hypre, parmetis and scotch to the corresponding AUR packages, removed sympy and added it to the list of dependencies from the official repos. Also the readme is updated accordingly. Pull request is on the way.
